core/store/migrations/migration1587580235/migrate.go

```diff
@@ -1,27 +1,21 @@
 package migration1587580235
 
 import (
-	"time"
-
-	"github.com/ethereum/go-ethereum/common"
 	"github.com/jinzhu/gorm"
-	"github.com/pkg/errors"
-	"github.com/smartcontractkit/chainlink/core/store/models"
 )
 
-type logConsumption struct {
-	ID           *models.ID  `gorm:"primary_key"`
-	BlockHash    common.Hash `gorm:"not null;unique_index:idx_unique_log_consumption"`
-	ConsumerType string      `gorm:"not null;unique_index:idx_unique_log_consumption"`
-	ConsumerID   *models.ID  `gorm:"not null;unique_index:idx_unique_log_consumption"`
-	LogIndex     uint        `gorm:"not null;unique_index:idx_unique_log_consumption"`
-	CreatedAt    time.Time
-}
-
-// Migrate adds the LogConsumption table
+// Migrate adds the LogConsumption table and adds a uniqueness constraint on the
+// combination of block_hash / log_index / consumer_id
 func Migrate(tx *gorm.DB) error {
-	if err := tx.AutoMigrate(&logConsumption{}).Error; err != nil {
-		return errors.Wrap(err, "could not add log_consumption table")
-	}
-	return nil
+	return tx.Exec(`
+	CREATE TABLE "log_consumptions" (
+		"id" serial primary key NOT NULL,
+		"block_hash" bytea NOT NULL,
+		"log_index" integer NOT NULL,
+		"consumer_id" uuid REFERENCES job_specs(id) ON DELETE CASCADE NOT NULL,
+		"created_at" timestamp without time zone DEFAULT now() NOT NULL
+	);
+
+	CREATE UNIQUE INDEX log_consumptions_block_hash_log_index_consumer_id_idx ON log_consumptions ("block_hash", "log_index", "consumer_id");
```

Comment:
Protip - ordering in multicolumn indexes makes a difference. If you put `consumer_id` first like so:

`("consumer_id", "block_hash", "log_index")`

we can use the index to look up log_consumptions by `consumer_id`, which is likely to be a common operationg. And we get it for free, because we had the index already anyway - nice!

In addtion, roughly how many log consumptions are we expecting?

If its likely to ever be more than 1000 or so, it might be worth adding a BRIN index on `created_at`.

Reply:
@sam that's a good question. We've been discussing adding a cleanup routine that runs once a day or something and deletes blocks older than X. For now though I would think it is extremely likely that this table grows to be larger than 1000.

Reply:
See story [here](https://www.pivotaltracker.com/story/show/172569628)